feat: add API to run application

atm1020 opened the pull request:
- Create a separate API to run applications.
-  example (with toogleterm):  
    ```lua 

    local runner = require('java').runner
    local Terminal = require('toggleterm.terminal').Terminal
    local CustomRunner = {}
    function CustomRunner.stop()
	    if CustomRunner.term then
		    CustomRunner.term:shutdown()
		    CustomRunner.term = nil
	    end
    end

    function CustomRunner.open_toogle_term(cmd)
	    local current_win = vim.api.nvim_get_current_win()
	    CustomRunner.term = Terminal:new({
		    cmd = table.concat(cmd, ' '),
		    hidden = true,
		    direction = 'horizontal',
		    close_on_exit = false,
		    float_opts = {},
		    on_open = function(_)
			    vim.api.nvim_set_current_win(current_win)
			    vim.cmd('stopinsert')
		    end,
	    })
	    CustomRunner.term:toggle()
    end
    
    
    vim.api.nvim_create_user_command('Run', function()
	    CustomRunner.stop()
	    runner.run_app(CustomRunner.open_toogle_term)
    end, {})
    
    vim.api.nvim_create_user_command('Stop', function()
	    CustomRunner.stop()
    end, {})

s1n7ax commented:
@atm1020 Since we have `nui` already I'm at the moment trying to figure out if we can open the terminal for user for the sake of completeness. I'm working on your branch so this PR may take some time.

atm1020 commented:
I started implementing a running option without using any plugins.
in nutshell:   I create a new buffer and set the log lines to that buffer from the background job, initiated with vim.fn.job.jobstart(...). 
Currently, it's in my neovim config, if you find it promising, im open to creating a draft pull request.

s1n7ax commented:
@atm1020 Sure. You can push changes to the same PR. All the work you are doing is going to be squashed commit in the end.

This is what I just tried. What ever works without additional dep would be fine.

```lua
local Split = require('nui.split')
local event = require('nui.utils.autocmd').event

local split = Split({
	relative = 'editor',
	position = 'bottom',
	size = '20%',
})

split:mount()

vim.fn.termopen('ls')
```

Requirements would be:
- User should be able to run the main method using `JavaRunnerRunMain` command
  - Additional arguments to `:JavaRunnerRunMain "hello world"` should be passed as application arguments
  - IF there are multiple main classes, user will be prompted to select the main method they want to run.
- A console should be opened (for the moment, bottom split would be fine I guess since it's very common in any IDE)
- On window close, window will be closed but application would still run in the background (when `hidden` option is true)
- User should be able to run `JavaRunnerViewLogs` and open the console back

atm1020 commented:
@s1n7ax 
I've extended the requirements a bit; I hope it's not a problem:
  - Instead of the `JavaRunnerViewLogs` function, I added the `JavaRunnerToogleLogs` functionality. So the user can open and close the log buffer using this feature.
  - I added `JavaRunnerStopMain` function.
 
Requirements question:
  - The `hidden` option is unclear to me. could you provide me with more information?

Additional idea:
 - I think it would be very useful to save a running profile for each project. This profile would include program arguments, allowing users to manage multiple configurations, and set defaults for the built in runner.

s1n7ax commented:
@atm1020 ignore the hide thing. It's probably working. I will test it. Profiling we could take separately. I wonder if something like overseer.nvim could help.

atm1020 commented:
@s1n7ax  Do you mean some kind of integration with overseer.nvim, or should it be used as an example ?

s1n7ax commented:
@atm1020 If you have completed `JavaRunnerToggleLogs` `JavaRunnerRunMain` `JavaRunnerStopMain` functions, for now lets use them. Right now program arguments should be able to be passed using the `JavaRunnerRunMain` command.

For instance `:JavaRunnerRunMain "helloworld"` would pass the argument to the application.

For creating profiles for projects, I think we should look into overseer to see if it could help. I will add a separate task to that.

s1n7ax commented:
Added following task to do runner profiles
https://github.com/nvim-java/nvim-java/issues/110

atm1020 commented:
@s1n7ax Yes, I have completed them, and the program arg option is working as well. I'm open to working on issue #110.

s1n7ax commented:
I will test this tomorrow

atm1020 commented:
@s1n7ax Im not sure if you have time to review it, but I have an additional idea to improve this feature (based on nvim-dap)  with sending the logs to the terminal instance instead of simply write a buffer.  For exampl with this change, the log outputs will be colorized. 
Can i push the changes to this branch?

s1n7ax commented:
@atm1020 for debug log you can create a separate Issue and PR. I have reviewed the current PR and added some comments.

atm1020 commented:
> @atm1020 for debug log you can create a separate Issue and PR. I have reviewed the current PR and added some comments.

@s1n7ax I don't see any comments on the pr. Am I missing something ?

atm1020 commented on lua/java/api/runner.lua:
```diff
@@ -0,0 +1,91 @@
+	local config_names = {}
+	local config_lookup = {}
+	for _, config in ipairs(configs) do
+		if config.projectName then
```
@s1n7ax  Yes, when I tested with a multimodule application, the `dap.get_dap_configs` method returned the following output
```
.nvim-java-core/lua/java-core/ls/clients/jdtls-client.lua:44: vscode.java.resolveMainClass success! response:  { {
    filePath = "/home/atm1020/workspace/maven-multi-module/module1/src/main/java/com/mycompany/maven/Module1Application.java",
    mainClass = "com.mycompany.maven.Module1Application",
    projectName = "module1"
  }, {
    filePath = "/home/atm1020/workspace/maven-multi-module/module2/src/main/java/com/mycompany/maven/Module2Application.java",
    mainClass = "com.mycompany.maven.Module2Application",
    projectName = "module2"
  }, {
    filePath = "/home/atm1020/workspace/maven-multi-module/module3/src/main/java/com/mycompany/maven/Module3Application.java",
    mainClass = "com.mycompany.maven.Module3Application",
    projectName = "module3"
  }, {
    filePath = "/home/atm1020/.cache/nvim/jdtls/workspaces/common-workspace/maven-multi-module/jdt.ls-java-project/src/com/mycompany/maven/Module2Application.java",
    mainClass = "com.mycompany.maven.Module2Application"
  }, {
    filePath = "/home/atm1020/.cache/nvim/jdtls/workspaces/common-workspace/maven-multi-module/jdt.ls-java-project/src/com/mycompany/maven/Module3Application.java",
    mainClass = "com.mycompany.maven.Module3Application"
  } }
```
Currently, I'm handling this issue by checking if the projectName is not nil.

 I forgot to create a bug ticket for this, i think it would be better to solve this in nvim-java-dap project

s1n7ax replied:
Do you have this sample java project pushed to a repo?

atm1020 replied:
I used this one: https://github.com/colinbut/maven-multi-module

atm1020 replied:
@s1n7ax Do you think this issue is valid? 
Should I remove the if statement or not?

s1n7ax replied:
@atm1020 No. It seems jdtls assigns some random project name in the config for single class files no?

```lua
{ {
    mainClass = "Test",
    name = "test_327a5bbc -> Test",
    projectName = "test_327a5bbc",
    request = "launch",
    type = "java"
  } }
```

atm1020 replied:
@s1n7ax  I think this relates to the jdtls cache. The "resolved mainClasses" always appear after a file from another module is opened.  Based on the logs, the project path points to the project cache folder, as you see in my previous comments (For example module2 and module3) .

atm1020 commented on lua/java/api/runner.lua:
```diff
@@ -0,0 +1,228 @@
+function BuiltInMainRunner:run_app(cmd)
+	self:stop(true)
+	if not self.bufnr then
+		self.bufnr = vim.api.nvim_create_buf(false, true)
```
Yes, you are right. I can solve this problem in issue  #115.

s1n7ax replied:
@atm1020 Wait, I thought you said logs when debugging. I thought the output buffer opened when DAP is started. You can make the changes to current PR then. My bad

atm1020 replied:
@s1n7ax  The output buffer will open when the application starts running, but currently it only writes logs to this buffer and its unable to handle user inputs. 
if I open a terminal instance in a buffer, it can handle stdin. 
Should I make this change in this pr? Sorry for the misunderstanding.

s1n7ax replied:
@atm1020 Yes use this PR

atm1020 commented on lua/java/api/runner.lua:
```diff
@@ -0,0 +1,228 @@
+	if self.is_open then
+		vim.api.nvim_buf_call(self.bufnr, function()
+			self:_set_up_buffer()
+			vim.cmd('normal! G')
```
Yes its working when i pass the last line number. 
```lua
local last_line = vim.fn.line("$")
vim.api.nvim_win_set_cursor(self.win, {last_line, 0})
```

Unfortunately, it is not working `-1` option, based on the manual, it should be.
```lua
vim.api.nvim_win_set_cursor(self.win, { -1, 0 })
```

Should I switch to using the `vim.api.nvim_win_set_cursor` function ?

s1n7ax replied:
Yes please. `nvim_win_set_cursor` is good. `norm G` would go to the last character anyway which probably the most annoying thing in the default behavior when looking at logs.

s1n7ax commented:
@atm1020 Since I have not heard from you I've decided to remove the PR and create a new PR with the changes. https://github.com/nvim-java/nvim-java/pull/122

atm1020 commented:
@s1n7ax Okey, I see it.  I've resolved all comments, so I thought you would merge or create a new change request if needed. Sorry for the misunderstanding!
